contracts/0.8.9/EIP712StETH.sol

```diff
@@ -1,27 +1,88 @@
+ *
+ * The encoding specified in the EIP is very generic, and such a generic implementation in Solidity is not feasible,
+ * thus this contract does not implement the encoding itself. Protocols need to implement the type-specific encoding
+ * they need in their contracts using a combination of `abi.encode` and `keccak256`.
+ *
+ * This contract implements the EIP 712 domain separator ({_domainSeparatorV4}) that is used as part of the encoding
+ * scheme, and the final step of the encoding to obtain the message digest that is then signed via ECDSA
+ * ({_hashTypedDataV4}).
+ *
+ * The implementation of the domain separator was designed to be as efficient as possible while still properly updating
+ * the chain id to protect against replay attacks on an eventual fork of the chain.
+ *
+ * NOTE: This contract implements the version of the encoding known as "v4", as implemented by the JSON RPC method
+ * https://docs.metamask.io/guide/signing-data.html[`eth_signTypedDataV4` in MetaMask].
+ *
+ */
+contract EIP712StETH is IEIP712 {
+    /* solhint-disable var-name-mixedcase */
+    // Cache the domain separator as an immutable value, but also store the chain id that it corresponds to, in order to
+    // invalidate the cached domain separator if the chain id changes.
+    bytes32 private immutable _CACHED_DOMAIN_SEPARATOR;
+    uint256 private immutable _CACHED_CHAIN_ID;
+    address private immutable _CACHED_STETH;
+
+    bytes32 private immutable _HASHED_NAME;
+    bytes32 private immutable _HASHED_VERSION;
+    bytes32 private immutable _TYPE_HASH;
+
+    error ZeroStETHAddress();
+
     /**
      * @dev Constructs specialized EIP712 instance for StETH token, version "2".
      */
-    constructor() EIP712("Liquid staked Ether 2.0", "2") {}
+    constructor(address _stETH) {
+        if (_stETH == address(0)) { revert ZeroStETHAddress(); }
+
+        bytes32 hashedName = keccak256(bytes("Liquid staked Ether 2.0"));
+        bytes32 hashedVersion = keccak256(bytes("2"));
+        bytes32 typeHash = keccak256(
+            "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+        );
+
+        _HASHED_NAME = hashedName;
+        _HASHED_VERSION = hashedVersion;
+        _CACHED_CHAIN_ID = block.chainid;
+        _CACHED_DOMAIN_SEPARATOR = _buildDomainSeparator(typeHash, hashedName, hashedVersion, _stETH);
+        _CACHED_STETH = _stETH;
+        _TYPE_HASH = typeHash;
+    }
 
     /**
      * @dev Returns the domain separator for the current chain.
      */
-    function domainSeparatorV4() external view override returns (bytes32) {
-        return _domainSeparatorV4();
+    function domainSeparatorV4(address _stETH) public view override returns (bytes32) {
+        if (_stETH == _CACHED_STETH && block.chainid == _CACHED_CHAIN_ID) {
+            return _CACHED_DOMAIN_SEPARATOR;
+        } else {
+            return _buildDomainSeparator(_TYPE_HASH, _HASHED_NAME, _HASHED_VERSION, _stETH);
```

Comment:
The idea is to prevent replay attack from other blockchains:
https://github.com/code-423n4/2021-09-swivel-findings/issues/98

and from the ERC-2612 spec: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2612.md
![image](https://user-images.githubusercontent.com/961317/220110051-6657f6fb-bca8-4337-8c97-5468691922b1.png)